File: java-workshop/part-3-mocks-clean/src/main/java/service/OrderService.java
package service;

import domain.Order;
import domain.PaymentResult;
import domain.PaymentStatus;
import domain.ShippingConfirmation;
import domain.StockReservation;
import repository.OrderRepository;

import java.util.List;

public class OrderService {
    private final OrderRepository orderRepository;
    private final StockService stockService;
    private final PricingService pricingService;
    private final PaymentService paymentService;
    private final ShippingService shippingService;

    public OrderService(
        OrderRepository orderRepository,
        StockService stockService,
        PricingService pricingService,
        PaymentService paymentService,
        ShippingService shippingService
    ) {
        this.orderRepository = orderRepository;
        this.stockService = stockService;
        this.pricingService = pricingService;
        this.paymentService = paymentService;
        this.shippingService = shippingService;
    }

    public OrderResult placeOrder(Order order) {
        List<StockReservation> reservations = stockService.reserveStock(order.items());
        if (reservations.stream().anyMatch(reservation -> !reservation.reserved())) {
            return new OrderResult.Failure(order.reject(), "Insufficient stock");
        }

        double totalPrice = pricingService.calculateTotal(order.items());
        PaymentResult paymentResult = paymentService.processPayment(
            order.id(),
            order.customerId(),
            totalPrice
        );

        if (paymentResult.status() != PaymentStatus.SUCCESS) {
            stockService.releaseStock(reservations);
            return new OrderResult.Failure(order.reject(), "Payment failed");
        }

        ShippingConfirmation shippingConfirmation = shippingService.createShipment(order);
        Order confirmedOrder = order.confirm(totalPrice);
        orderRepository.save(confirmedOrder);
        return new OrderResult.Success(confirmedOrder, shippingConfirmation);
    }
}

File: kotlin-workshop/part-1-powermock/src/main/kotlin/service/OrderService.kt
package service

import domain.Order
import domain.OrderStatus
import domain.PaymentStatus
import domain.ShippingConfirmation
import infrastructure.Database

/**
 * Main orchestrator service - creates all its dependencies internally
 * This is the "PowerMock nightmare" - to test this in isolation,
 * you need to mock constructors, objects, and singletons
 */
class OrderService {
    
    fun createOrder(order: Order): OrderResult {
        // Create dependencies internally (bad practice!)
        val stockService = StockService()
        val pricingService = PricingService()
        val paymentService = PaymentService()
        val shippingService = ShippingService()
        
        // 1. Check stock availability
        if (!stockService.checkAvailability(order.items)) {
            return OrderResult.Failure(order.reject(), "Insufficient stock")
        }
        
        // 2. Calculate total price
        val totalPrice = pricingService.calculateTotal(order.items)
        
        // 3. Process payment
        val paymentResult = paymentService.processPayment(
            order.id, 
            order.customerId, 
            totalPrice
        )
        
        if (paymentResult.status != PaymentStatus.SUCCESS) {
            return OrderResult.Failure(order.reject(), "Payment failed")
        }
        
        // 4. Reserve stock
        val reservations = stockService.reserveStock(order.items)
        if (reservations.any { !it.reserved }) {
            // Rollback payment
            paymentResult.transactionId?.let { paymentService.refundPayment(it) }
            stockService.releaseStock(reservations)
            return OrderResult.Failure(order.reject(), "Could not reserve stock")
        }
        
        // 5. Create shipment
        val shippingConfirmation = shippingService.createShipment(order)
        
        // 6. Save to database and return confirmed order
        val confirmedOrder = order.confirm(totalPrice)
        Database.saveOrder(confirmedOrder.id, mapOf(
            "status" to confirmedOrder.status,
            "totalPrice" to totalPrice,
            "trackingNumber" to shippingConfirmation.trackingNumber
        ))
        
        return OrderResult.Success(confirmedOrder, shippingConfirmation)
    }
}

sealed class OrderResult {
    data class Success(
        val order: Order,
        val shippingConfirmation: ShippingConfirmation
    ) : OrderResult()
    
    data class Failure(
        val order: Order,
        val reason: String
    ) : OrderResult()
}

